lib/Sema/TypeCheckAccess.cpp

```diff
@@ -969,69 +972,89 @@ void UsableFromInlineChecker::check(Decl *D) {
     auto PBD = cast<PatternBindingDecl>(D);
     bool isTypeContext = PBD->getDeclContext()->isTypeContext();
 
+    // Stored instance properties in public/@usableFromInline fixed-contents
+    // structs in resilient modules must always use public/@usableFromInline
+    // types. In these cases, check the access against the struct instead of the
+    // VarDecl, and customize the diagnostics. (The code below doesn't check
+    // for "in resilient modules" because there's no reason to write
+    // @_fixedLayout on a struct in a non-resilient module.)
+    const ValueDecl *fixedLayoutStructContext = nullptr;
+    if (auto *parentStruct = dyn_cast<StructDecl>(PBD->getDeclContext())) {
+      if (parentStruct->getAttrs().hasAttribute<FixedLayoutAttr>() &&
+          !PBD->isStatic() && PBD->hasStorage()) {
+        fixedLayoutStructContext = parentStruct;
+      }
+    }
+
     llvm::DenseSet<const VarDecl *> seenVars;
-    for (auto entry : PBD->getPatternList())
-    entry.getPattern()->forEachNode([&](const Pattern *P) {
-      if (auto *NP = dyn_cast<NamedPattern>(P)) {
-        // Only check individual variables if we didn't check an enclosing
-        // TypedPattern.
-        const VarDecl *theVar = NP->getDecl();
-        if (theVar->getFormalAccess() != AccessLevel::Internal)
+    for (auto entry : PBD->getPatternList()) {
+      entry.getPattern()->forEachNode([&](const Pattern *P) {
+        if (auto *NP = dyn_cast<NamedPattern>(P)) {
+          // Only check individual variables if we didn't check an enclosing
+          // TypedPattern.
+          const VarDecl *theVar = NP->getDecl();
+          if (!fixedLayoutStructContext && shouldSkipChecking(theVar))
+            return;
+          if (seenVars.count(theVar) || theVar->isInvalid())
+            return;
+
+          checkTypeAccess(theVar->getType(), nullptr,
```

Comment:
nitpick: prefer getInterfaceType() to avoid creating archetypes (I know it was here already)